swift/extractor/SwiftOutputRewrite.cpp

```diff
@@ -0,0 +1,318 @@
+    llvm::StringRef path(oldPath);
+    auto aliases = computeModuleAliases(path, targetTriple);
+    for (auto& alias : aliases) {
+      remapping[alias] = newPath;
+    }
+  }
+
+  return remapping;
+}
+
+void ensureNewPathsExist(const std::unordered_map<std::string, std::string>& remapping) {
+  for (auto& [_, newPath] : remapping) {
+    llvm::SmallString<PATH_MAX> filepath(newPath);
+    llvm::StringRef parent = llvm::sys::path::parent_path(filepath);
+    if (std::error_code ec = llvm::sys::fs::create_directories(parent)) {
+      std::cerr << "Cannot create redirected directory: " << ec.message() << "\n";
+    }
+  }
+}
+
+void storeRemappingForVFS(const SwiftExtractorConfiguration& config,
+                          const std::unordered_map<std::string, std::string>& remapping) {
+  // Only create remapping for the .swiftmodule files
+  std::unordered_map<std::string, std::string> modules;
+  for (auto& [oldPath, newPath] : remapping) {
+    if (llvm::StringRef(oldPath).endswith(".swiftmodule")) {
+      modules[oldPath] = newPath;
+    }
+  }
+
+  if (modules.empty()) {
+    return;
+  }
+
+  if (std::error_code ec = llvm::sys::fs::create_directories(config.tempVFSDir)) {
+    std::cerr << "Cannot create temp VFS directory: " << ec.message() << "\n";
+    return;
+  }
+
+  if (std::error_code ec = llvm::sys::fs::create_directories(config.VFSDir)) {
+    std::cerr << "Cannot create VFS directory: " << ec.message() << "\n";
+    return;
+  }
+
+  // Constructing the VFS yaml file in a temp folder so that the other process doesn't read it
+  // while it is not complete
+  // TODO: Pick a more robust way to not collide with files from other processes
+  auto tempVfsPath = config.tempVFSDir + '/' + std::to_string(getpid()) + "-vfs.yaml";
+  std::error_code ec;
+  llvm::raw_fd_ostream fd(tempVfsPath, ec, llvm::sys::fs::OF_None);
+  if (ec) {
+    std::cerr << "Cannot create temp VFS file: '" << tempVfsPath << "': " << ec.message() << "\n";
+    return;
+  }
+  // TODO: there must be a better API than this
+  // LLVM expects the version to be 0
+  fd << "{ version: 0,\n";
+  // This tells the FS not to fallback to the physical file system in case the remapped file is not
+  // present
+  fd << "  fallthrough: false,\n";
+  fd << "  roots: [\n";
+  for (auto& [oldPath, newPath] : modules) {
+    fd << "  {\n";
+    fd << "    type: 'file',\n";
+    fd << "    name: '" << oldPath << "\',\n";
+    fd << "    external-contents: '" << newPath << "\'\n";
+    fd << "  },\n";
+  }
+  fd << "  ]\n";
+  fd << "}\n";
+
+  fd.flush();
+  auto vfsPath = config.VFSDir + '/' + std::to_string(getpid()) + "-vfs.yaml";
+  if (std::error_code ec = llvm::sys::fs::rename(tempVfsPath, vfsPath)) {
+    std::cerr << "Cannot move temp VFS file '" << tempVfsPath << "' -> '" << vfsPath
+              << "': " << ec.message() << "\n";
+    return;
+  }
+}
+
+std::vector<std::string> collectVFSFiles(const SwiftExtractorConfiguration& config) {
```

Comment:
I assume this is done separately from `storeRemappingForVFS`, because we might be trying to find a file from a previous extractor invocation? If so, that might be worth a comment.

Reply:
Yeah, it's a "collective knowledge" here for the processes that run later.